spices/SPICE-0004-HTTP-Proxy-Support.adoc

```diff
@@ -0,0 +1,180 @@
+// suppress inspection "HttpUrlsUsage" for whole file
+= HTTP Proxy Support
+
+* Proposal: link:./SPICE-0004-HTTP-Proxy-Support.adoc[SPICE-0004]
+* Author: link:https://github.com/bioball[Dan Chao]
+* Status: TBD
+* Implemented in: TBD
+* Category: Language, Standard Library
+
+== Introduction
+
+This SPICE describes support for HTTP proxying.
+
+== Motivation
+
+Some environments do not allow direct connections to the internet, instead requiring that connections go through an intermediary proxy.
+
+To enable Pkl in these situations, this SPICE provides a way to redirect outbound requests through said proxies.
+
+== Proposed Solution
+
+Pkl will support HTTP proxies.
+Proxy settings will be added, and this will control how Pkl makes HTTP requests.
+
+== Detailed design
+
+=== Configuring proxies
+
+A proxy can be set via the following, in order of precedence (least to greatest):
+
+1. `~/.pkl/settings` file
+2. `PklProject` file (in `evaluatorSettings`)
+3. CLI Flags/pkl-gradle options/`HttpClient#Builder`/server mode message passing API
+
+==== Environment Variables
+
+Proxies will _not_ be configured via environment variables.
+This maintains consistency with the rest of Pkl; it does not support using environment variables for any other configuration.
+Additionally, while other languages might use environment variables, they have slightly different meaning from one language to another (see link:https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy[]]).
+Not using environment variables avoids a possble problem where the same environment variable influences both Pkl and another runtime, but in different ways.
+
+==== Settings file
+
+A new `proxy` property is added to the global settings file.
+
+[source,pkl]
+----
+module pkl.settings
+
+import "pkl:Proxy"
+
+/// HTTP proxy settings.
+@Since { version = "0.26.0" }
+proxy: Proxy?
+----
+
+==== `PklProject`
+
+A new `proxy` property is added to `EvaluatorSettings` of a PklProject file.
+
+[source,pkl]
+----
+module pkl.Project
+
+import "pkl:Proxy"
+
+class EvaluatorSettings {
+  proxy: Proxy?
+}
+----
+
+[[new-stdlib-module]]
+==== New stdlib module: `pkl.Proxy`
```

Comment:
Having a stdlib module just for proxy settings feels strange.
Can’t this be part of pkl.settings, even if pkl.Project needs it too (which I’m not convinced of)?

Reply:
Both `pkl.settings` and `pkl.Project` need this, and neither is a logical subset of the other, so it feels cleanest to put it in a separate module.

We can also put it in `base.pkl`, but that affects the global namespace, which doesn't feel great.

We also possibly should put redirect rules in there, so maybe this intermediary module should be called `HttpSettings`, with proxy and rewrite rules both being in there.

Reply:
If a new stdlib module is needed, its scope should definitely be wider than proxy settings.